File: java/Easy/_383_RansomNote.java
package Easy;

import org.testng.annotations.Test;

public class _383_RansomNote {
    public boolean canConstruct(String ransomNote, String magazine) {
        char[] rs = ransomNote.toCharArray();
        char[] mgz = magazine.toCharArray();
        int cnt = 0;
        for(int i = 0; i<rs.length; i++){
            for(int j = 0; j<mgz.length; j++){
                if(mgz[j] == rs[i]){
                    cnt++;
                    mgz[j] = ' ';
                    break;
                }
            }
        }
        
        return cnt == rs.length;
    }

    public boolean canConstruct2(String ransomNote, String magazine) {
        int[] c = new int[26];
        char[] rs = ransomNote.toCharArray();
        char[] mgz = magazine.toCharArray();
        for(char r : mgz){
            c[r-'a']++;
        }

        for(int i = 0; i<rs.length; i++){
            if(--c[rs[i]-'a'] < 0) return false;
        }
        return true;
    }

    @Test
    public void test(){
        System.out.println(canConstruct2("aa", "aab"));
    }
}

File: Python3/_383_RansomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        dic = {}
        for m in magazine:
            if m in dic.keys():
                dic[m] += 1
            else:
                dic[m] = 1
        for r in ransomNote:
            if r in dic.keys():
                dic[r] -= 1
            else:
                return False

        for key in dic.keys():
            if dic[key] < 0:
                return False
        return True
test = Solution()
test.canConstruct("ab","ab")
